cirq/sim/mux.py

```diff
@@ -251,11 +264,14 @@ def final_density_matrix(
         operations to the desired initial state.
 
     """
-    initial_state_like = None
-    if not isinstance(initial_state, int):
-        initial_state_like = np.asarray(initial_state, dtype=dtype)
+    qubit_order = ops.QubitOrder.as_qubit_order(qubit_order)
+
+    if isinstance(initial_state, int):
+        pass
+    elif isinstance(initial_state, value.TensorProductState):
```

Comment:
This logic is liable to repeated a bunch of times. Consider extracting a method like `initial_state_like_to_state_vector`.

Reply:
Luckily this now exists (`cirq.to_valid_state_vector`). Will move everything there.